cmd/lncli/cmd_pay.go

```diff
@@ -467,20 +454,113 @@ func trackPayment(ctx *cli.Context) error {
 		return err
 	}
 
+	_, err = printLivePayment(stream)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// printLivePayment uses advanced terminal control codes to rewrite the output
+// with the latest payment state. This continues until the payment reaches a
+// final state.
+func printLivePayment(stream routerrpc.Router_TrackPaymentClient) (
+	*lnrpc.Payment, error) {
+
+	writer := uilive.New()
+	writer.Start()
+	defer writer.Stop()
 	for {
-		status, err := stream.Recv()
+		payment, err := stream.Recv()
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		printRespJSON(status)
+		table := formatPayment(payment)
+		writer.Write([]byte(table))
 
-		if status.Status != lnrpc.Payment_IN_FLIGHT {
-			return nil
+		if payment.Status != lnrpc.Payment_IN_FLIGHT {
+			return payment, nil
 		}
 	}
 }
 
+// formatPayment formats the payment state as an ascii table.
+func formatPayment(payment *lnrpc.Payment) string {
+	t := table.NewWriter()
+
+	// Build table header.
+	t.AppendHeader(table.Row{
+		"HTLC_STATE", "ATTEMPT_TIME", "RESOLVE_TIME", "RECEIVER_AMT",
+		"FEE", "TIMELOCK", "CHAN_OUT", "ROUTE",
+	})
+	t.SetColumnConfigs([]table.ColumnConfig{
+		table.ColumnConfig{Name: "ATTEMPT_TIME", Align: text.AlignRight},
+		table.ColumnConfig{Name: "RESOLVE_TIME", Align: text.AlignRight},
+		table.ColumnConfig{Name: "CHAN_OUT", Align: text.AlignLeft,
+			AlignHeader: text.AlignLeft},
+	})
+
+	// Add all htlcs as rows.
+	createTime := time.Unix(0, payment.CreationTimeNs)
+	for _, htlc := range payment.Htlcs {
+		attemptTime := time.Unix(0, htlc.AttemptTimeNs)
+		attemptMs := attemptTime.Sub(createTime).Milliseconds()
+		attemptTimeText := fmt.Sprintf("%.3f", float64(attemptMs)/1000.0)
+
+		resolveTimeText := "-"
+		if htlc.ResolveTimeNs != 0 {
+			resolveTime := time.Unix(0, htlc.ResolveTimeNs)
+			resolveTimeNs := resolveTime.Sub(createTime).Milliseconds()
+			resolveTimeText = fmt.Sprintf(
+				"%.3f", float64(resolveTimeNs)/1000.0,
+			)
+		}
+
+		route := htlc.Route
+		lastHop := route.Hops[len(route.Hops)-1]
+
+		hops := []string{}
+		for _, h := range route.Hops {
+			hops = append(hops, fmt.Sprintf("%v", h.PubKey[:6]))
+		}
+
+		state := htlc.Status.String()
+		if htlc.Failure != nil {
+			state = fmt.Sprintf(
+				"%v @ %v",
+				htlc.Failure.Code,
+				htlc.Failure.FailureSourceIndex,
+			)
+		}
+
+		t.AppendRow([]interface{}{
+			state, attemptTimeText, resolveTimeText,
+			lastHop.AmtToForwardMsat, route.TotalFeesMsat,
+			route.TotalTimeLock, route.Hops[0].ChanId,
+			strings.Join(hops, "->")},
+		)
+	}
+
+	// Render table.
+	b := &bytes.Buffer{}
+	t.SetOutputMirror(b)
+	t.Render()
+
+	// Add additional payment-level data.
+	fmt.Fprintf(b, "Payment status: %v", payment.Status)
```

Comment:
would be nice to also log some more info here, like:
```
Destination:    lightning-merchant (or full pubkey)
Amount + Fee:   10000000 + 1010 mSat
Payment Hash:   713cb852db70a11ddf128695f0824055924531b8f56e3f31422321eb7690e254
Payment Status: SUCCEEDED, preimage: 89b4a95726520c3c3e9048da2808a33421337d24a0ebe39033d576df71444bfa
```
the fee could be updated live as the sum of the total fee for all non-failed htlc attempts

Reply:
Added except for the destination. Would need to re-decode the payment request for that. Or inspect any of the routes last hops, but there may be none.

This also becomes more of a ux discussion. I am fine with leaving it as is now. The most important thing, readable htlcs, is there. Further improvements can be made for 0.11, also depending on what users want to see.

Reply:
> Would need to re-decode the payment request for that. Or inspect any of the routes last hops, but there may be none.

We already [decode the invoice when there is one](https://github.com/lightningnetwork/lnd/blob/master/cmd/lncli/cmd_pay.go#L361), is there a way for no destination to be specified when there is no invoice? Or is there another reason we'd need to re-decode?

> This also becomes more of a ux discussion. 

This is true. Only reason I pointed these out here is because it seemed logical to at least print info we already do during confirmation, but it's not a blocker.

Reply:
If there was a confirmation step, the info is also printed above the table. If there was no confirmation step, the info is on the command line. Only when `payinvoice -f` is used, the info is only shown in the htlc rows. 

I don't want to say that this isn't a good suggestion. But if more adaptations and re-review makes this miss the cut, I'd rather leave it as is.

Reply:
sgtm, we can leave as is and revisit later